jax/_src/pallas/core.py

```diff
@@ -591,149 +619,151 @@ def __init__(
-            index_map_avals=index_map_avals,
-            index_map_tree=index_map_tree,
-            grid=grid_mapping_grid,
-            mapped_dims=(),
-            what="outputs",
-        ),
-        flat_out_specs,
-        out_paths,
-        out_avals,
-    )
-    grid_mapping = GridMapping(
-        grid=grid_mapping_grid,  # type: ignore[arg-type]
-        grid_names=self.grid_names,
-        block_mappings=(*in_block_mappings, *out_block_mappings),
-        index_map_avals=index_map_avals,  # type: ignore[arg-type]
-        index_map_tree=index_map_tree,
-        vmapped_dims=(),
-        num_index_operands=num_flat_scalar_prefetch,
-        num_constant_operands=0,  # Fixed up later
-        num_inputs=len(flat_in_specs),
-        num_outputs=len(flat_out_specs),
-        num_scratch_operands=num_flat_scratch_operands,
-    )
-    grid_mapping.check_invariants()
-    in_ref_avals = [bm.block_aval for bm in in_block_mappings]
-    jaxpr_in_ref_avals = tree_util.tree_unflatten(in_tree, in_ref_avals)
-    jaxpr_in_avals = (*jaxpr_scalar_ref_avals, *jaxpr_in_ref_avals)
-    out_ref_avals = [bm.block_aval for bm in out_block_mappings]
-    jaxpr_out_avals = tree_util.tree_unflatten(out_tree, out_ref_avals)
-    if not isinstance(jaxpr_out_avals, (tuple, list)):
-      jaxpr_out_avals = (jaxpr_out_avals,)
-    return (*jaxpr_in_avals, *jaxpr_out_avals,
-            *jaxpr_scratch_avals), grid_mapping
-
   def _make_scratch_aval(self, obj: object) -> jax_core.AbstractValue:
     assert False  # Not needed in GridSpec
 
   def _make_scalar_ref_aval(self, aval):
     assert False  # Not needed in GridSpec
 
-  def unzip_dynamic_grid_bounds(
-      self,
-  ) -> tuple[GridSpec, tuple[Any, ...]]:
-    static_grid = tuple(
-        d if isinstance(d, int) else None for d in self.grid
-    )
-    dynamic_bounds = tuple(d for d in self.grid if not isinstance(d, int))
-    # We can't use dataclasses.replace, because our fields are incompatible
-    # with __init__'s signature.
-    static_self = copy.copy(self)
-    static_self.grid = static_grid  # type: ignore
-    return static_self, dynamic_bounds
+
+def get_grid_mapping(
+    grid_spec: GridSpec,
+    in_avals: Sequence[jax_core.AbstractValue],
+    in_tree: tree_util.PyTreeDef,
+    in_paths: Sequence[tree_util.KeyPath],
+    out_avals: Sequence[jax_core.AbstractValue],
+    out_tree: tree_util.PyTreeDef,
+    out_paths: Sequence[tree_util.KeyPath],
+) -> tuple[tuple[jax_core.AbstractValue, ...],
+           GridMapping]:
+  assert all(i is None or isinstance(i, int) for i in grid_spec.grid)
+  grid_mapping_grid = tuple(
+      dynamic_grid_dim if d is None else d for d in grid_spec.grid
+  )
+  # The inputs for the index maps
+  index_map_avals = (
+      (index_map_grid_aval,) * len(grid_spec.grid))
+  index_map_tree = tree_util.tree_structure((index_map_avals, {}))
+
+  num_scalar_prefetch: int = getattr(grid_spec, "num_scalar_prefetch", 0)
```

Comment:
It's a bit unsatisfying that Pallas core has to know about `num_scalar_prefetch` and `scratch_shapes`, since both are TPU-specific.

I can't think of a better way, but just flagging in case you have ideas.

Reply:
I am actually thinking that we can merge the functionality of PrefetchScalarGridSpec from the TPU into the core. 

But even if we don't do that, I think it is worth having a single calling convention, with optional parts that may be applicable only to some platform.